python/pyspark/sql/types.py

```diff
@@ -186,6 +186,30 @@ def fromInternal(self, ts):
             return datetime.datetime.fromtimestamp(ts // 1000000).replace(microsecond=ts % 1000000)
 
 
+class CalendarIntervalType(DataType, metaclass=DataTypeSingleton):
```

Comment:
There have been a lot of discussions about exposing interval type in other language APIs but I lost the track. @yaooqinn and @cloud-fan, are we going to make internal as a proper exposed type? Or only support it in some contexts?

Reply:
Yes, I glanced over a few threads and couldn't really figure out where it is going, hence a limited scope of this PR. However, if the type is supported in multiple contexts, current behavior doesn't seem like an intended one.

Reply:
Doesn't @zero323's example from the PR description show that Spark already exposes this type? 

```python
spark.sql("SELECT current_date() - current_date()")
```

For the record, btw, Postgres supports [an `interval` type](https://www.postgresql.org/docs/current/datatype-datetime.html) and done so since at least [version 7.1](https://www.postgresql.org/docs/7.1/datatype-datetime.html), which was released in 2001.

Reply:
The problem is that it has been half-exposed so far. There have been many discussions up to which context we should support. e.g.) `CalendarInterval` is marked as `Unstable`.

Reply:
BTW, I personally agree with adding this in particular due to @nchammas point https://github.com/apache/spark/pull/29935#discussion_r562175621 here